src/artifacts-cred.py

```diff
@@ -0,0 +1,177 @@
+                "netfx",
+                "CredentialProvider.Microsoft",
+                "CredentialProvider.Microsoft.exe",
+            )
+            self.exe = [tool_path]
+        else:
+            try:
+                sys_version = tuple(int(i) for i in
+                    subprocess.check_output(["dotnet", "--version"]).decode().strip().partition("-")[0].split("."))
+                get_runtime_path = lambda: "dotnet"
+            except Exception as e:
+                message = (
+                    "Unable to find dependency dotnet, please manually install"
+                    " the .NET SDK and ensure 'dotnet' is in your PATH. Error: "
+                )
+                raise Exception(message + str(e))
+
+            #TODO: fix the path
+            tool_path = os.path.join(
+                os.path.dirname(os.path.abspath(__file__)),
+                "plugins",
+                "plugins",
+                "netcore",
+                "CredentialProvider.Microsoft",
+                "CredentialProvider.Microsoft.dll",
+            )
+            self.exe = [get_runtime_path(), "exec", tool_path]
+
+        if not os.path.exists(tool_path):
+            raise RuntimeError("Unable to find credential provider in the expected path: " + tool_path)
+
+
+    def get_credentials(self, url):
+        # Public feed short circuit: return nothing if the endpoint is public (can authenticate without credentials).
+        if self._can_authenticate(url, None):
+            return None, None
+
+        # Getting credentials with IsRetry=false; the credentials may come from the cache
+        username, password = self._get_credentials_from_credential_provider(url, is_retry=False)
+        
+        # Do not attempt to validate if the credentials could not be obtained
+        if username is None or password is None:
+            return username, password
+
+        # Make sure the credentials are still valid (i.e. not expired)
+        if self._can_authenticate(url, (username, password)):
+            return username, password
+
+        # The cached credentials are expired; get fresh ones with IsRetry=true
+        return self._get_credentials_from_credential_provider(url, is_retry=True)
+
+    def _can_authenticate(self, url, auth):
+        response = requests.get(url, auth=auth)
+
+        return response.status_code < 500 and \
+            response.status_code != 401 and \
+            response.status_code != 403
+
+
+    def _get_credentials_from_credential_provider(self, url, is_retry):
+        non_interactive = self._NON_INTERACTIVE_VAR_NAME in os.environ and \
+            os.environ[self._NON_INTERACTIVE_VAR_NAME] and \
+            str(os.environ[self._NON_INTERACTIVE_VAR_NAME]).lower() == "true"
+
+        proc = Popen(
+            self.exe + [
+                "-Uri", url,
+                "-IsRetry", str(is_retry),
+                "-NonInteractive", str(non_interactive),
+                "-CanShowDialog", "False",
```

Comment:
-CanShowDialog False inhibits the best interactive prompts. We don't want to fall back to Device Code unless we have to.

```suggestion
                "-CanShowDialog", str(not non_interactive),
```

Reply:
Ideally we (or artifacts-credprovider) would determine the _correct_ value for this, but it's not trivial

Reply:
I see. For my understanding, why is it default to false for artifacts keyring?